### PR TITLE
Add checksum to snapshot metadata files

### LINKED ISSUE #11589: Snapshot/Restore: Add checksums to metadata files in repository
Currently, we only checksums of the data files, leaving possibility for metadata files to get corrupted without us noticing it. We should add checksums to the metadata files and verify them on restore.

### PR DESCRIPTION
**Pull request by imotov**
This commit adds checksum to snapshot files that store global and index based metadata as well as shard information.

Closes #11589

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        return new FsBlobStore(settings, randomRepoPath());
+    }
+
+    protected void randomCorruption(BlobContainer blobContainer, String blobName) throws IOException {
+        byte[] buffer = new byte[(int)blobContainer.listBlobsByPrefix(blobName).get(blobName).length()];
+        try (InputStream inputStream = blobContainer.openInput(blobName)) {
+            Streams.readFully(inputStream, buffer);
+        }
+        int location = randomIntBetween(0, buffer.length - 1);
+        buffer[location] = (byte)(buffer[location] ^ 42);
```
Comment (rmuir) [inline core/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java]: can we compute CRC on `buffer` before and after introducing this corrupted byte. We can assume() the checksum has actually changed, so the test does not fail rarely?

**Comment by imotov** [comment]
@rmuir I pushed the fix for the checksum issue. Could you or @s1monw review when you have a chance?

File: core/src/main/java/org/elasticsearch/repositories/blobstore/AtomicSnapshotBlobStoreFormat.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import org.elasticsearch.common.blobstore.BlobContainer;
+import org.elasticsearch.common.xcontent.XContentType;
+
+import java.io.IOException;
+
+/**
+ * A version of {@link BlobStoreFormat} that can automatically resolve blob names base on snapshot id
+ * and writes blobs in atomic manner
+ */
+public abstract class AtomicSnapshotBlobStoreFormat<T> extends  SnapshotBlobStoreFormat<T> {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/AtomicSnapshotBlobStoreFormat.java]: can we just fold this into thr default and always make the move?

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreFormat.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+     * @param compress true if the content should be compressed
+     * @throws IOException
+     */
+    public void write(T obj, BlobContainer blobContainer, String blobName, boolean compress) throws IOException {
+        BytesReference bytes = write(obj, compress);
+        try (OutputStream outputStream = blobContainer.createOutput(blobName)) {
+            if (checksum) {
+                try (StreamOutput streamOutput = new OutputStreamStreamOutput(outputStream)) {
+                    try (BufferedChecksumStreamOutput checksumStreamOutput = new BufferedChecksumStreamOutput(streamOutput)) {
+                        checksumStreamOutput.writeLong(BLOB_STORE_MAGIC);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreFormat.java]: can we use CodecUtils to write the header and footer?

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreFormat.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+     * @param blobName blob name
+     * @return parsed blob object
+     * @throws IOException
+     */
+    public T read(BlobContainer blobContainer, String blobName) throws IOException {
+        try (InputStream inputStream = blobContainer.openInput(blobName)) {
+            if (checksum) {
+                try(StreamInput streamInput = new InputStreamStreamInput(inputStream)) {
+                    try (BufferedChecksumStreamInput checksumStreamInput = new BufferedChecksumStreamInput(streamInput)) {
+                        long magic = checksumStreamInput.readLong();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreFormat.java]: can we use CodecUtils for this too?

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreFormat.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+                }
+            } else {
+                write(obj, bytesStreamOutput);
+            }
+            return bytesStreamOutput.bytes();
+        }
+    }
+
+    private void write(T obj, StreamOutput streamOutput) throws IOException {
+        XContentBuilder builder = XContentFactory.contentBuilder(xContentType, streamOutput);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreFormat.java]: can we use try with here?

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 6]
```diff
@@ -155,6 +136,125 @@
...
+        }
+
+        @Override
+        protected MetaData fromXContent(XContentParser parser) throws IOException {
+            XContentParser.Token token;
+            if ((token = parser.nextToken()) == XContentParser.Token.FIELD_NAME) {
+                parser.nextToken();
+                return MetaData.Builder.fromXContent(parser);
+            }
+            throw new ElasticsearchParseException("unexpected token  [" + token + "]");
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: we can maybe use similar technique as we do in `QueryParsingException` and also report the location

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 7]
```diff
@@ -155,6 +136,125 @@
...
+        }
+    }
+
+    private final SnapshotBlobStoreFormat<MetaData> globalMetaDataFormat = new GlobalMetaDataFormat(false) {
+        @Override
+        protected String blobName(SnapshotId snapshotId) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: instead of subclassing all this can we just use some kind of String format for this that we pass as a ctor arg. and for the legacy / non-legacy case we create them as we need?

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 6 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Can I address this as a separate PR? There are a lot of ElasticsearchParseExceptions in the snapshot/restore code and this PR is already big enough.

**Comment by imotov** [comment]
@s1monw I pushed the changes as we discussed and opened #12105 to track parsing exception changes.

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+     * Writes properly formatted blob in atomic manner.
+     * <p/>
+     * The blob will be compressed and checksum will be written if required.
+     *
+     * @param obj           object to be serialized
+     * @param blobContainer blob container
+     * @param name          blob name
+     * @throws IOException
+     */
+    public void write(T obj, BlobContainer blobContainer, String name) throws IOException {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: I think we should use `writeAtomic` everywhere just to reduce the complexity.

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Locale;
+
+/**
+ * Snapshot metadata file format used in v2.0 and above
+ */
+public class ChecksumBlobStoreFormat<T extends ToXContent> extends BlobStoreFormat<T> {
+
+    public static final String TEMP_FILE_PREFIX = "pending-";
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: Can this be private? It should be an implementation detail of this class only I think?

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+
+    private static final XContentType DEFAULT_X_CONTENT_TYPE = XContentType.JSON;
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: We should change this to smile

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+
+    // The format version
+    public static final int VERSION = 1;
+
+    private static final int BUFFER_SIZE = 4096;
+
+    protected final XContentType xContentType;
+
+    protected final boolean wrapper;
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: Can you add a comment for what this boolean is?

File: core/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+import org.apache.lucene.store.IndexInput;
+
+import java.io.EOFException;
+import java.io.IOException;
+
+/**
+ * Wraps array of bytes into IndexInput
+ */
+public class ByteArrayIndexInput extends IndexInput {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java]: Can you add some unit tests for this class?

File: core/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    private final byte[] bytes;
+
+    private int pos;
+
+    private int offset;
+
+    private int length;
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java]: As an aside, maybe we should consider changing our APIs to use `long` so we don't have to worry about byte arrays larger than ~2.14gb?

File: core/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+    public ByteArrayIndexInput(String resourceDesc, byte[] bytes, int offset, int length) {
+        super(resourceDesc);
+        this.bytes = bytes;
+        this.offset = offset;
+        this.length = length;
+    }
+
+    @Override
+    public void close() throws IOException {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java]: I think this should call `super.close()`?

File: core/src/main/java/org/elasticsearch/common/xcontent/FromXContentBuilder.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import org.elasticsearch.common.ParseFieldMatcher;
+
+import java.io.IOException;
+
+/**
+ *
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/common/xcontent/FromXContentBuilder.java]: Needs some javadoc :)

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreFormat.java [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+
+    protected final FromXContentBuilder<T> reader;
+
+    protected final ParseFieldMatcher parseFieldMatcher;
+
+    // Serialization parameters to specify correct context for metadata serialization
+    protected static final ToXContent.Params SNAPSHOT_ONLY_FORMAT_PARAMS;
+
+    static {
+        Map<String, String> snpashotOnlyParams = Maps.newHashMap();
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreFormat.java]: snpashot -> snapshot

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreFormat.java [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+        snpashotOnlyParams.put(MetaData.CONTEXT_MODE_PARAM, MetaData.CONTEXT_MODE_SNAPSHOT);
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreFormat.java]: It's unclear what exactly is going on here, can you add a comment as to why this is needed and what it does?

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 18]
```diff
@@ -129,23 +116,34 @@
...
     private static final String TESTS_FILE = "tests-";
 
-    private static final String METADATA_PREFIX = "meta-";
+    private static final String METADATA_NAME_FORMAT = "meta-%s.dat";
 
-    private static final String LEGACY_METADATA_PREFIX = "metadata-";
+    private static final String LEGACY_METADATA_NAME_FORMAT = "metadata-%s";
 
-    private static final String METADATA_SUFFIX = ".dat";
+    private static final String METADATA_CODEC = "meta-data";
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Can this be "metadata"? We don't hyphenate anywhere else in the code.

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 19]
```diff
@@ -129,23 +116,34 @@
...
 
-    private final BlobStoreIndexShardRepository indexShardRepository;
+    private static final String INDEX_METADATA_CODEC = "index-meta-data";
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Same here, maybe "index-metadata"

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 20]
```diff
@@ -325,7 +329,7 @@ public void deleteSnapshot(SnapshotId snapshotId) {
                 BlobPath indexPath = basePath().add("indices").add(index);
                 BlobContainer indexMetaDataBlobContainer = blobStore().blobContainer(indexPath);
                 try {
-                    indexMetaDataBlobContainer.deleteBlob(blobName);
+                    indexMetaDataFormat(snapshot.version()).delete(indexMetaDataBlobContainer, snapshotId.getSnapshot());
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Do we need to delete the `indexMetaDataLegacyFormat` somewhere as well?

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 21]
```diff
@@ -506,125 +493,36 @@ private RateLimiter getRateLimiter(RepositorySettings repositorySettings, String
...
-                if ((token = parser.nextToken()) == XContentParser.Token.FIELD_NAME) {
-                    parser.nextToken();
-                    MetaData metaData = MetaData.Builder.fromXContent(parser);
-                    if ((token = parser.nextToken()) == XContentParser.Token.END_OBJECT) {
-                        return metaData;
-                    }
-                }
-            }
-            throw new ElasticsearchParseException("unexpected token [{}]", token);
+    private BlobStoreFormat<MetaData> globalMetaDataFormat(Version version) {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Need some javadoc here :)

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 22]
```diff
@@ -506,125 +493,36 @@ private RateLimiter getRateLimiter(RepositorySettings repositorySettings, String
...
-     * Returns name of metadata blob
-     *
-     * @param snapshotId snapshot id
-     * @param legacy true if legacy (pre-2.0.0) format should be used
-     * @return name of metadata blob
-     */
-    private String metaDataBlobName(SnapshotId snapshotId, boolean legacy) {
-        if (legacy) {
-            return LEGACY_METADATA_PREFIX + snapshotId.getSnapshot();
+    private BlobStoreFormat<Snapshot> snapshotFormat(Version version) {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Same here.

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 23]
```diff
@@ -506,125 +493,36 @@ private RateLimiter getRateLimiter(RepositorySettings repositorySettings, String
...
-     * @return BytesStreamOutput representing JSON serialized global MetaData
-     * @throws IOException
-     */
-    private void writeGlobalMetaData(MetaData metaData, StreamOutput outputStream) throws IOException {
-        XContentBuilder builder = XContentFactory.contentBuilder(XContentType.JSON, outputStream);
-        builder.startObject();
-        MetaData.Builder.toXContent(metaData, builder, snapshotOnlyFormatParams);
-        builder.endObject();
-        builder.flush();
+    private BlobStoreFormat<IndexMetaData> indexMetaDataFormat(Version version) {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: And here, you removed all the Javadoc! :)

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java (hunk 11 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: Oh I see it below, makes sense, `wrapper` just sounds like a noun instead of a flag

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+    public T readBlob(BlobContainer blobContainer, String blobName) throws IOException {
+        try (InputStream inputStream = blobContainer.openInput(blobName)) {
+            byte[] bytes = ByteStreams.toByteArray(inputStream);
+            final String resourceDesc = "ChecksumBlobStoreFormat.readBlob(blob=\"" + blobName + "\")";
+            try (ByteArrayIndexInput indexInput = new ByteArrayIndexInput(resourceDesc, bytes)) {
+                CodecUtil.checksumEntireFile(indexInput);
+                CodecUtil.checkHeader(indexInput, codec, VERSION, VERSION);
+                long filePointer = indexInput.getFilePointer();
+                long contentSize = indexInput.length() - CodecUtil.footerLength() - filePointer;
+                BytesReference bytesReference = new BytesArray(bytes, (int)filePointer, (int)contentSize);
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: Yikes if we end up truncating documents due to the long->int cast here!

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+     * @param blobContainer blob container
+     * @param name          blob name
+     * @throws IOException
+     */
+    public void write(T obj, BlobContainer blobContainer, String name) throws IOException {
+        String blobName = blobName(name);
+        writeBlob(obj, blobContainer, blobName);
+    }
+
+    public void writeBlob(T obj, BlobContainer blobContainer, String blobName) throws IOException {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: Should this be private/protected?

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 26]
```diff
@@ -0,0 +1,189 @@
...
+        BytesReference bytes = write(obj);
+        try (OutputStream outputStream = blobContainer.createOutput(blobName)) {
+            final String resourceDesc = "ChecksumBlobStoreFormat.writeBlob(blob=\"" + blobName + "\")";
+            try (OutputStreamIndexOutput indexOutput = new OutputStreamIndexOutput(resourceDesc, outputStream, BUFFER_SIZE)) {
+                CodecUtil.writeHeader(indexOutput, codec, VERSION);
+                try (OutputStream indexOutputOutputStream = new org.elasticsearch.common.lucene.store.OutputStreamIndexOutput(indexOutput) {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: fully-qualified name shouldn't be needed here?

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+            } else {
+                write(obj, bytesStreamOutput);
+            }
+            return bytesStreamOutput.bytes();
+        }
+    }
+
+    protected void write(T obj, StreamOutput streamOutput) throws IOException {
+        try (XContentBuilder builder = XContentFactory.contentBuilder(xContentType, streamOutput)) {
+            if (wrapper) {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: Maybe we can add a comment here (or where `wrapper` is defined) about why it's necessary?

File: core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreTests.java [hunk 28]
```diff
@@ -90,6 +90,7 @@ public void basicWorkFlowTest() throws Exception {
                 .setType("fs").setSettings(Settings.settingsBuilder()
                         .put("location", randomRepoPath())
                         .put("compress", randomBoolean())
+                        .put("compress", false)
```
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreTests.java]: This doesn't make sense with the above setting? Randomize and then hardcode false?

**Comment by dakrone** [comment]
Left some comments

File: core/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java (hunk 13 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java]: The IndexInput API is using `long`. But here I am switching to `int` because I am dealing with a byte array and array indices have type `int` and it's not possible to create an array with size specified by `long` value. Something like this `new byte[3L]` will not even compile. In practice, the maximum array size is even smaller. Depending on java VM implementation but it can be between `Integer.MAX_VALUE-2` and `Integer.MAX_VALUE-5`. So, I think we should be fine in terms of both theoretical and practical limits but after your comment I have updated all comparisons in ByteArrayIndexInput to use `long` type to makes sure we don't miss integer overflow. Thanks!

File: core/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java (hunk 14 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java]: It cannot because `super.close()` is abstract.

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 20 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: When we deal with indices here, we know the version of the snapshot that created these indices. So `indexMetaDataFormat(snapshot.version())` here gives us the proper format (`indexMetaDataLegacyFormat` for old snapshots or `indexMetaDataFormat` for new snapshots) that we delete.

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 23 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: It's a different method. So I didn't remove javadoc, I just didn't add it :)

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java (hunk 24 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: That's not really possible. See my comments about byte arrays above. If file is too big, it would fail a few lines before when we try to read it:
```
byte[] bytes = ByteStreams.toByteArray(inputStream);
```

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java (hunk 26 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: These are two different class with the same name. The first `OutputStreamIndexOutput ` 3 lines above is `org.apache.lucene.store.OutputStreamIndexOutput`. It is `IndexOutput` facade that writes into `OutputStream`. 

The second `OutputStreamIndexOutput` with fully-qualified name is `OutputStream` facade that writes into `IndexOutput`. 

All this makes me thing that `org.elasticsearch.common.lucene.store.OutputStreamIndexOutput` should be renamed to `IndexOutputOutputStream`. I will do that.

File: core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreTests.java (hunk 28 above)
Comment (imotov) [inline core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreTests.java]: It shouldn't make sense - it's debugging leftovers. Will remove.

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java (hunk 8 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: I tried to do that but it creates a lot of inefficiency on slow repositories and problems while trying to create a snapshot with the same name as previously failed snapshot. We can do this switch, but I would prefer move it into another PR.

File: core/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java (hunk 12 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/common/lucene/store/ByteArrayIndexInput.java]: Added the unit test.

**Comment by imotov** [comment]
@dakrone thanks for the review. I pushed fixes.

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 29]
```diff
@@ -0,0 +1,217 @@
...
+    // The format version
+    public static final int VERSION = 1;
+
+    private static final int BUFFER_SIZE = 4096;
+
+    protected final XContentType xContentType;
+
+    protected final boolean compress;
+
+    public final String codec;
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: does this need to be public?

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java [hunk 30]
```diff
@@ -0,0 +1,217 @@
...
+     */
+    public void writeAtomic(T obj, BlobContainer blobContainer, String name) throws IOException {
+        String blobName = blobName(name);
+        String tempBlobName = tempBlobName(name);
+        writeBlob(obj, blobContainer, tempBlobName);
+        try {
+            blobContainer.move(tempBlobName, blobName);
+        } catch (IOException ex) {
+            // Move failed - try cleaning up
+            blobContainer.deleteBlob(tempBlobName);
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: Maybe we should inject a logger so we can log something here? Otherwise we silently fail the writeAtomic?

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java (hunk 8 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: Makes sense

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java (hunk 30 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: Or maybe we need to rethrow the exception?

File: core/src/main/java/org/elasticsearch/snapshots/Snapshot.java [hunk 31]
```diff
@@ -277,9 +293,14 @@ public static Snapshot fromXContent(XContentParser parser) throws IOException {
...
-        XContentParser.Token token = parser.currentToken();
-        if (token == XContentParser.Token.START_OBJECT) {
+        if (parser.currentToken() == null) { // fresh parser? move to the first token
+            parser.nextToken();
+        }
+        if (parser.currentToken() == XContentParser.Token.START_OBJECT) {  // on a start object move to next token
+            parser.nextToken();
+        }
+        XContentParser.Token token;
+        if ((token = parser.nextToken()) == XContentParser.Token.START_OBJECT) {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/snapshots/Snapshot.java]: I don't quite understand this logic, we are looking for a double-nested object? like `{ { ... } }`?

**Comment by dakrone** [comment]
Left a couple more comments, other than that I have a question, I see you creating and passing around a `ParseFieldMatcher` in the classes, but not using it anywhere, am I missing where you are using it?

File: core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java (hunk 30 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java]: Yeah, I forgot to rethrow. That was definitely the intention. Great catch!

File: core/src/main/java/org/elasticsearch/snapshots/Snapshot.java (hunk 31 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/snapshots/Snapshot.java]: No, we are looking for `{ "snapshot": { .... } }`. First we make sure that we came from the fresh parser, during this check the current token is null. So we set it to the first `{`. The second if verifies that we have got this `{` and moves parser once step further. So, now the current token is field name `snapshot`. The third if moves parser to the second `{` with `parser.currentName()` still being `snapshot`. I hope it makes sense. This structure enables us to use this parser in case we just opened a new file and in case we found it inside  a bigger XContent like
```
{
    "some_parent": {
        "something": {
        },
        "snapshot": {
           ......
        }
    }
}
```

**Comment by imotov** [comment]
@dakrone I pushed the revision. Currently, `ParseFieldMatcher ` is used in `BlobStoreIndexShardSnapshot`, `BlobStoreIndexShardSnapshots`. I could probably get rid of it in these classes but `ParseFieldMatcher`'s usage will be probably expanded in the future. So, it's nice to keep it handy.

**Comment by dakrone** [comment]
LGTM
